File: src/lib/governance/investigation.ts
import { db } from "@/lib/db";
import { logAction } from "@/lib/audit";

// Ported from aegis's api/routers/investigation.py — breach investigation
// detail, restore, and confirm-breach, remapped onto the unified Case model
// (sourceType AGENT_GOVERNANCE, sourceRef = agent.id) instead of aegis's
// separate BreachInvestigation table.

export async function getAgentInvestigation(agentId: string) {
  const agent = await db.agent.findUnique({ where: { id: agentId } });
  if (!agent) return null;

  const kase = await db.case.findFirst({
    where: { sourceType: "AGENT_GOVERNANCE", sourceRef: agentId },
    orderBy: { createdAt: "desc" },
    include: {
      detections: { orderBy: { createdAt: "desc" } },
      timeline: { orderBy: { occurredAt: "asc" } },
    },
  });

  return {
    agent,
    case: kase,
    recommendation: kase ? recommendationFor(kase.severity) : null,
  };
}

function recommendationFor(severity: string) {
  if (severity === "CRITICAL") {
    return "RECOMMENDED: Keep revoked. Issue a new identity with stricter policies.";
  }
  if (severity === "HIGH") {
    return "RECOMMENDED: Investigate thoroughly before restoring. Update security policies.";
  }
  return "RECOMMENDED: Review evidence. May be a false positive — consider restoration.";
}

export async function restoreAgent(agentId: string, justification: string, actorId?: string) {
  const agent = await db.agent.findUniqueOrThrow({ where: { id: agentId } });
  if (agent.status !== "REVOKED") {
    throw new Error("Agent is not revoked");
  }

  await db.agent.update({
    where: { id: agentId },
    data: { status: "ACTIVE", trust: 100, level: 10, mode: "FULL_ACCESS" },
  });

  const kase = await db.case.findFirst({
    where: { sourceType: "AGENT_GOVERNANCE", sourceRef: agentId },
    orderBy: { createdAt: "desc" },
  });
  if (kase) {
    await db.case.update({
      where: { id: kase.id },
      data: { status: "FALSE_POSITIVE", resolvedAt: new Date() },
    });
  }

  await logAction({
    actorId,
    caseId: kase?.id,
    action: "governance.agent_restored",
    detail: { agentId, justification },
  });
}

export async function confirmBreach(agentId: string, notes: string, actorId?: string) {
  const kase = await db.case.findFirst({
    where: { sourceType: "AGENT_GOVERNANCE", sourceRef: agentId },
    orderBy: { createdAt: "desc" },
  });
  if (!kase) throw new Error("No investigation found for this agent");

  await db.case.update({
    where: { id: kase.id },
    data: { status: "RESOLVED", resolvedAt: new Date() },
  });

  await logAction({
    actorId,
    caseId: kase.id,
    action: "governance.breach_confirmed",
    detail: { agentId, notes },
  });
}

File: legacy/components/aegis/backend/src/aegis/api/routers/investigation.py
from fastapi import APIRouter, HTTPException
from sqlmodel import Session, select
from aegis.db import engine, Agent, BreachInvestigation
from typing import Dict, Any
import json
import time

router = APIRouter(tags=["Investigation"])

@router.get("/agents/{agent_id}/investigation")
def get_investigation(agent_id: str):
    """Get detailed breach investigation for a revoked agent"""
    with Session(engine) as session:
        # Get agent
        agent = session.get(Agent, agent_id)
        if not agent:
            raise HTTPException(status_code=404, detail="Agent not found")
        
        # Get investigation
        investigation = session.exec(
            select(BreachInvestigation).where(BreachInvestigation.agent_id == agent_id)
        ).first()
        
        if not investigation:
            raise HTTPException(status_code=404, detail="No investigation found for this agent")
        
        # Parse evidence JSON
        evidence = json.loads(investigation.evidence) if investigation.evidence else {}
        
        return {
            "agent_id": agent_id,
            "agent_status": agent.status,
            "trust": agent.trust,
            "breach_type": investigation.breach_type,
            "severity": investigation.severity,
            "detection_mechanism": investigation.detection_mechanism,
            "evidence": evidence,
            "timestamp": investigation.timestamp,
            "revoked_at": investigation.revoked_at,
            "status": investigation.status,
            "recommendation": get_recommendation(investigation)
        }

@router.post("/agents/{agent_id}/restore")
def restore_agent(agent_id: str, justification: str = ""):
    """Restore a revoked agent after investigation"""
    with Session(engine) as session:
        # Get agent
        agent = session.get(Agent, agent_id)
        if not agent:
            raise HTTPException(status_code=404, detail="Agent not found")
        
        if agent.status != "REVOKED":
            raise HTTPException(status_code=400, detail="Agent is not revoked")
        
        # Update agent
        agent.status = "Active"
        agent.trust = 100.0
        agent.level = 10
        agent.mode = "FULL_ACCESS"
        session.add(agent)
        
        # Update investigation
        investigation = session.exec(
            select(BreachInvestigation).where(BreachInvestigation.agent_id == agent_id)
        ).first()
        
        if investigation:
            investigation.status = "RESTORED"
            session.add(investigation)
        
        session.commit()
        
        return {
            "status": "success",
            "message": f"Agent {agent_id} restored",
            "justification": justification
        }

@router.post("/agents/{agent_id}/confirm-breach")
def confirm_breach(agent_id: str, notes: str = ""):
    """Confirm a breach investigation and permanently revoke agent"""
    with Session(engine) as session:
        investigation = session.exec(
            select(BreachInvestigation).where(BreachInvestigation.agent_id == agent_id)
        ).first()
        
        if not investigation:
            raise HTTPException(status_code=404, detail="No investigation found")
        
        investigation.status = "CONFIRMED"
        session.add(investigation)
        session.commit()
        
        return {
            "status": "success",
            "message": f"Breach confirmed for {agent_id}",
            "notes": notes
        }

def get_recommendation(investigation: BreachInvestigation) -> str:
    """Get recommended action based on investigation"""
    if investigation.severity == "CRITICAL":
        return "RECOMMENDED: Keep revoked. Issue new identity with stricter policies."
    elif investigation.severity == "HIGH":
        return "RECOMMENDED: Investigate thoroughly before restoring. Update security policies."
    else:
        return "RECOMMENDED: Review evidence. May be false positive - consider restoration."
